synapse/storage/databases/main/deviceinbox.py

```diff
@@ -474,10 +479,14 @@ async def delete_messages_for_device(
 
         def delete_messages_for_device_txn(txn: LoggingTransaction) -> int:
             sql = (
-                "DELETE FROM device_inbox"
-                " WHERE user_id = ? AND device_id = ?"
-                " AND stream_id <= ?"
+                f"DELETE FROM device_inbox WHERE {self.database_engine.row_id_name} IN ("
+                f"SELECT {self.database_engine.row_id_name} FROM device_inbox"
+                " WHERE user_id = ? AND device_id = ? AND stream_id <= ?"
             )
+            if limit:
+                sql += f" LIMIT {limit}"
```

Comment:
I think we only call this function in one place, so can we make `limit` non-optional to simplify things?

Reply:
It's also called in the [sync handler](https://github.com/matrix-org/synapse/blob/develop/synapse/handlers/sync.py#L363).
I think I am also going to switch this call to using the scheduler too, it should improve sync perf quite a bit when the inbox is big.